plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -176,67 +188,71 @@ public DeleteResult delete() throws IOException {
     }
 
     @Override
-    public void deleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException {
+    public void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) throws IOException {
         doDeleteBlobs(blobNames, true);
     }
 
-    private void doDeleteBlobs(List<String> blobNames, boolean relative) throws IOException {
-        if (blobNames.isEmpty()) {
+    private void doDeleteBlobs(Iterator<String> blobNames, boolean relative) throws IOException {
+        if (blobNames.hasNext() == false) {
             return;
         }
-        final Set<String> outstanding;
+        final Iterator<String> outstanding;
```

Comment:
The blobs are distinct as a result of the logic upstream. The reason we used the set was to track what is still outstanding, iterating the set was kind of by accident (we don't do anything of the kind for other implementations either and were just iterating over the list directly).
I don't think this matters too much and it would be somewhat painful to add an assertion here as you point out.

Reply:
:+1: